### ISSUE TITLE
What should be trained, and Which parts use pretrained models?

### ISSUE THREAD

**Issue by y0ngw00** [neu-vi/OmniControl#15]
Hi, I have some questions on the implementation level. Sorry for so many questions :rofl: 

(1) **About CMDM**
- Are the weights of positional encoding and Transformer encoder of Realism guidance control from MDM pretrained model, or should I train them with weight initialization?
- Should I use the embedding step of CMDM separately from that of MDM?
- The control is the output of zero_convs layer, and I don't understand the role of zero_convs. Is the zero_convs just a zero-weight layers and trained thru back-propagation? I'm not sure what it is.
- For the realism guidance, guided_hint is not only added to noised_input, but it is also a control that is added to the latents in the middle of layers. So it is added twice. Am I right?
![image](https://github.com/neu-vi/OmniControl/assets/48562460/3a22108d-a3dd-446d-9b39-d63b1b378440)
![image](https://github.com/neu-vi/OmniControl/assets/48562460/51f85807-e83c-4f5f-9594-56510bbdb4c7)

(2) **Spatial guidance**
- If I'd like to use this model with other datasets, It seems me to prepare some presets: raw_mean, raw_std, and hint blocks.
raw_mean and raw_std seems mean and std of global position of joints. But I don't know the representations of hint blocks. Are they zero-mean normalized? or relative the pelvis?
- It seems like the spatial guidance is applied before the diffusion time step. It is confusing because I think it is different from the paper(spatial guidance after each diffusion step). Is the method changed? or the same as the architecture in the paper? 
```
x_t = self.q_sample(x_start, t, noise=noise)
x_t = self.guide(x_t, t, model_kwargs=model_kwargs, train=True)
terms = {}
model_output = model(x_t, self._scale_timesteps(t), **model_kwargs)
```

(3) etc
- What is the role of input_hint_block? Why the poseEmbedding of input_hint_block has initialized zero-weights in the last layer?

**Comment by ymingxie** [comment]
Sorry for the late response. 
1. Yes during the training the weight of Realism guidance should be initialized from MDM. 
2. Yeah the embedding step in MDM and CMDM should be separate. 
3. Zero-conv module is initialized with zero weight and zero bias, so the cmdm won't affect the mdm performance at the beginning of the training. See [ControlNet](https://github.com/lllyasviel/ControlNet) for more details. 
4. Actually it's added several times: the control features are added in all layers of MDM. See [ControlNet](https://github.com/lllyasviel/ControlNet) for more details: In ControlNet the control features are added to corresponding layers in the UNet Encoder. In our case we only have encoder so we added to all layers.
5. The hint_block is a learnable module. To be specific, it contains several mlp layers to encode the raw representation of human motion to latent space.
6. In the training stage we run spatial guidance first because we want the realism guidance to learn to amend the edited motion by spatial guidance. In the inference stage, the spatial guidance is after the realism guidance (same as the paper describes).

**Comment by y0ngw00** [comment]
Well understood. Thank you 👍
